federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,762 @@
+	// max above we abruptly start removing replicas from some/all clusters.
+	// Here we might make some max and min to 0, signalling that this hpa be
+	// removed from this cluster altogether
+	for remainingMax < 0 {
+		for _, replicas := range scheduled {
+			if replicas != nil &&
+				!((replicas.max - replicas.min) < 0) {
+				replicas.max--
+				remainingMax++
+				if remainingMax == 0 {
+					break
+				}
+			}
+		}
+	}
+
+	return remainingMax
+}
+
+func distributeMaxReplicas(toDistributeMax int32, lists hpaLists, rdc replicaNums,
+	currentObjs map[string]pkgruntime.Object, scheduled map[string]*replicaNums) int32 {
+	for cluster := range scheduled {
+		if toDistributeMax == 0 {
+			break
+		}
+		if scheduled[cluster] == nil {
+			continue
+		}
+		if maxReplicasNeeded(currentObjs[cluster]) {
+			scheduled[cluster].max++
+			if lists.availableMax.Len() > 0 {
+				popped, notEmpty := lists.availableMax.PopAny()
+				if notEmpty {
+					// Boundary checks have happened earlier in
+					// minReplicasReducible().
+					scheduled[popped].max--
+				}
+			}
+			// Any which ways utilise available map replicas
+			toDistributeMax--
+		}
+	}
+
+	// If we have new clusters where we can  give our replicas,
+	// then give away all our replicas to the new clusters first.
+	if lists.noHpa.Len() > 0 {
+		for toDistributeMax > 0 {
+			for _, cluster := range lists.noHpa.UnsortedList() {
+				if scheduled[cluster] == nil {
+					scheduled[cluster] = &replicaNums{min: 0, max: 0}
+				}
+				// first give away max from clusters offering them
+				// this case especially helps getting hpa into newly joining
+				// clusters.
+				if lists.availableMax.Len() > 0 {
+					popped, notEmpty := lists.availableMax.PopAny()
+					if notEmpty {
+						// Boundary checks to reduce max have happened earlier in
+						// minReplicasReducible().
+						scheduled[cluster].max++
+						scheduled[popped].max--
+						toDistributeMax--
+						continue
+					}
+				}
+				if toDistributeMax < rdc.max {
+					scheduled[cluster].max += toDistributeMax
+					toDistributeMax = 0
+					break
+				}
+				scheduled[cluster].max += rdc.max
+				toDistributeMax -= rdc.max
+			}
+		}
+	} else { // we have no new clusters but if still have max replicas to distribute;
+		// just distribute all in current clusters.
+		for toDistributeMax > 0 {
+			for cluster, replicas := range scheduled {
+				if replicas == nil {
+					replicas = &replicaNums{min: 0, max: 0}
+					scheduled[cluster] = replicas
+				}
+				// First give away max from clusters offering them.
+				// This case especially helps getting hpa into newly joining
+				// clusters.
+				if lists.availableMax.Len() > 0 {
+					popped, notEmpty := lists.availableMax.PopAny()
+					if notEmpty {
+						// Boundary checks have happened earlier in
+						// minReplicasReducible().
+						replicas.max++
+						scheduled[popped].max--
+						toDistributeMax--
+						continue
+					}
+				}
+				if toDistributeMax < rdc.max {
```

Comment:
Again here you dump all remaining replicas on one cluster.  I don't understand why.  Can you explain?  It seems that you should distribute the remaining replicas across all clusters, rather than fill one cluster, and leave the others empty?